### PR TITLE
Nodejs week1

### PR DESCRIPTION
**Pull request by halynabondar**
Homework Node.js week1

### CODE REVIEW

File: api/src/index.js [hunk 1]
```diff
@@ -5,12 +5,77 @@ import bodyParser from "body-parser";
...
 const apiRouter = express.Router();
 
+app.get("/my-route", (req, res) => {
+  res.send("Hi friend");
+});
+
+// Respond with all meals in the future (relative to the 'when' datetime)
+app.get('/future-meals', async (req, res) => {
+  try {
+    const meals = await knex.raw("SELECT * FROM Meal WHERE `when` > NOW()");
```
Comment (urbanogilson) [inline api/src/index.js]: The recommended approach to writing queries is to use [Knex Query Builder](https://knexjs.org/guide/query-builder.html)

File: api/src/index.js [hunk 2]
```diff
@@ -5,12 +5,77 @@ import bodyParser from "body-parser";
...
+    res.json(meals[0].length ? meals[0] : []); // Return empty array if no meals
+  } catch (err) {
+    res.status(500).json({ error: 'Internal server error' });
+  }
+});
+
+// Respond with all meals sorted by ID
+app.get('/all-meals', async (req, res) => {
+  try {
+    const meals = await knex.raw("SELECT * FROM Meal ORDER BY id");
```
Comment (urbanogilson) [inline api/src/index.js]: Using [Knex Query Builder](https://knexjs.org/guide/query-builder.html) we could simplify this to:
```suggestion
    const meals = await knex("Meal").orderBy("id");
```

File: api/src/index.js [hunk 3]
```diff
@@ -5,12 +5,77 @@ import bodyParser from "body-parser";
 import knex from "./database_client.js";
 import nestedRouter from "./routers/nested.js";
 
+const meals = await knex.raw("SELECT * FROM Meal");
+console.log(meals);
+
```
Comment (urbanogilson) [inline api/src/index.js]: Remove temporary code

File: api/src/index.js [hunk 4]
```diff
@@ -5,12 +5,77 @@ import bodyParser from "body-parser";
...
+    res.json(meals[0].length ? meals[0] : []); // meals[0] because knex.raw returns an array
+  } catch (err) {
+    res.status(500).json({ error: 'Internal server error' });
+  }
+});
+
+// Respond with all meals in the past (relative to the 'when' datetime)
+app.get('/past-meals', async (req, res) => {
+  try {
+    const meals = await knex.raw("SELECT * FROM Meal WHERE `when` < NOW()");
```
Comment (urbanogilson) [inline api/src/index.js]: You named your table Meal, checkout this page for SQL table/column name conventions: [https://brainstation.io/learn/sql/naming-conventions](https://brainstation.io/learn/sql/naming-conventions)

**Comment by urbanogilson** [comment]
Good job! See my inline comment

File: api/src/index.js [hunk 5]
```diff
@@ -3,29 +3,77 @@ import express from "express";
...
+  res.send("Hi friend");
 });
 
-// This nested router example can also be replaced with your own sub-router
-apiRouter.use("/nested", nestedRouter);
+// Respond with all meals in the future (relative to the 'when' datetime)
+app.get('/future-meals', async (req, res) => {
+  try {
+    const meals = await knex("meal").where("when", ">", knex.fn.now());
+    res.json(meals[0].length ? meals[0] : []); // meals[0] because knex.raw returns an array
```
Comment (urbanogilson) [inline api/src/index.js]: The requirement is to return all meals in the future
> /future-meals	Respond with all meals in the future (relative to the when datetime)

```suggestion
    res.json(meals);
```

File: api/src/index.js [hunk 6]
```diff
@@ -3,29 +3,77 @@ import express from "express";
...
+  } catch (err) {
+    res.status(500).json({ error: 'Internal server error' });
+  }
+});
+
+// Respond with all meals in the past (relative to the 'when' datetime)
+app.get('/past-meals', async (req, res) => {
+  try {
+    const meals = await knex("meal").where("when", "<", knex.fn.now());
+    res.json(meals[0].length ? meals[0] : []); // Return empty array if no meals
```
Comment (urbanogilson) [inline api/src/index.js]: > /past-meals	Respond with **all meals i**n the past (relative to the when datetime)

File: api/src/index.js [hunk 7]
```diff
@@ -3,29 +3,77 @@ import express from "express";
...
+  } catch (err) {
+    res.status(500).json({ error: 'Internal server error' });
+  }
+});
+
+// Respond with all meals sorted by ID
+app.get('/all-meals', async (req, res) => {
+  try {
+    const meals = await knex("Meal").orderBy("id");
+    res.json(meals[0].length ? meals[0] : []);
```
Comment (urbanogilson) [inline api/src/index.js]: > /all-meals	Respond with **all meals** sorted by ID

```suggestion
    res.json(meals);
```

File: api/src/index.js [hunk 8]
```diff
@@ -3,29 +3,77 @@ import express from "express";
...
+  } catch (err) {
+    res.status(500).json({ error: 'Internal server error' });
+  }
+});
+
+// Respond with the first meal (meaning with the minimum id)
+app.get('/first-meal', async (req, res) => {
+  try {
+    const meal = await knex("meal").orderBy("id").limit(1);
```
Comment (urbanogilson) [inline api/src/index.js]: Leverage the query builder to make the code more readable, e.g. https://knexjs.org/guide/schema-builder.html#first

```suggestion
    const meal = await knex("meal").orderBy("id").first();
```

File: api/src/index.js [hunk 9]
```diff
@@ -3,29 +3,77 @@ import express from "express";
...
+    if (meal[0].length) {
+      res.json(meal[0][0]); // Return the first meal object
```
Comment (urbanogilson) [inline api/src/index.js]:
```suggestion
    if (meal) {
      res.json(meal); // Return the first meal object
```

File: api/src/index.js [hunk 10]
```diff
@@ -3,29 +3,77 @@ import express from "express";
...
+    }
+  } catch (err) {
+    res.status(500).json({ error: 'Internal server error' });
+  }
+});
+
+// Respond with the last meal (meaning with the maximum id)
+app.get('/last-meal', async (req, res) => {
+  try {
+    const meal = await knex("meal").orderBy("id", "desc").limit(1);
```
Comment (urbanogilson) [inline api/src/index.js]:
```suggestion
    const meal = await knex("meal").orderBy("id", "desc").first();
```

File: api/src/index.js [hunk 11]
```diff
@@ -3,29 +3,77 @@ import express from "express";
...
+    if (meal[0].length) {
+      res.json(meal[0][0]); // Return the last meal object
```
Comment (urbanogilson) [inline api/src/index.js]:
```suggestion
    if (meal) {
      res.json(meal); // Return the last meal object
```

File: api/src/index.js (hunk 5 above)
Comment (halynabondar) [reply api/src/index.js]: Thank you very much for the tips

File: api/src/index.js [hunk 12]
```diff
@@ -3,29 +3,79 @@ import express from "express";
...
+  } catch (err) {
+    res.status(500).json({ error: 'Internal server error' });
+  }
+});
+
+// Respond with the first meal (meaning with the minimum id)
+app.get('/first-meal', async (req, res) => {
+  try {
+    const meal = await knex("meal").orderBy("id").first();
+    res.json(meal[0][0]); // Return the first meal object
```
Comment (urbanogilson) [inline api/src/index.js]:
```suggestion

```

File: api/src/index.js [hunk 13]
```diff
@@ -3,29 +3,79 @@ import express from "express";
...
+  } catch (err) {
+    res.status(500).json({ error: 'Internal server error' });
+  }
+});
+
+// Respond with the last meal (meaning with the maximum id)
+app.get('/last-meal', async (req, res) => {
+  try {
+    const meal = await knex("meal").orderBy("id", "desc").first();
+    res.json(meal[0][0]); // Return the last meal object
```
Comment (urbanogilson) [inline api/src/index.js]:
```suggestion
```

**Pull request merged**